File: scala/src/main/scala/0005_LongestPalindromicSubstring.scala
object LongestPalindromicSubstring extends App {
  /*
    "babad"

    all substrings: n^2
    isPanlindrome: n

    O(n^3), O(1)

    isPanlidrome : n
    all of substring: n^2

     l
    "babad"
     |


    Determining panlindrome from middle.

    isPalinderome: N
    all possible substirng: N

    O(n^2) O(1)

   */
  def longestPalindrome(s: String): String = {

    def findPalindrome(l: Int, r: Int): (Int, Int) = {
      if (l < 0 || r >= s.length()) (l + 1, r - 1)
      else if (s.charAt(l) == s.charAt(r)) findPalindrome(l - 1, r + 1)
      else (l + 1, r - 1)
    }

    def go(i: Int, start: Int, to: Int): String = {
      if (i >= s.length()) s.substring(start, to + 1)
      else {
        val (x, y) = List(
          (start, to),
          findPalindrome(i, i + 1), // even
          findPalindrome(i, i) // odd
        ).maxBy { case (x1, y2) => y2 - x1 }

        go(i + 1, x, y)
      }
    }

    go(0, 0, 0)
  }

  assert(longestPalindrome("babad") == "bab")
  assert(longestPalindrome("cbbd") == "bb")
}

File: rust/src/bin/0005_longest_palindromic_substring.rs
fn main() {
    assert_eq!(longest_palindrome("babad".to_string()), "bab");
    assert_eq!(longest_palindrome("cbbd".to_string()), "bb");
}

/*
 *
 * "babad"
 *
 * brute force:
 * is_palindrome: O(n)
 *  i ->
 * "babad"
 *   <- j
 * all substrings: O(n^2)
 *  i ->
 * "babad"
 *  j ->
 *
 * O(n^3), O(1)
 *
 * determining it from center.
 * moving center: n
 * find a palindrome: n
 *
 * O(n^2), O(1)
 *
 */
pub fn longest_palindrome(s: String) -> String {
    let mut start = 0;
    let mut end = 0;

    let ss = s.as_bytes();
    for center in 0..ss.len() {
        // for odd length
        for i in 0..=center.min(ss.len() - 1 - center) {
            let j = center - i;
            let k = center + i;

            if ss[j] != ss[k] {
                break;
            } else if k - j > end - start {
                start = j;
                end = k;
            }
        }

        // for even length
        for i in 0..=center.min(ss.len() - 1 - center) {
            let j = center - i;
            let k = center + i + 1;

            if k >= ss.len() || ss[j] != ss[k] {
                break;
            } else if k - j > end - start {
                start = j;
                end = k;
            }
        }
    }

    String::from_utf8(ss[start..=end].to_owned()).unwrap()
}
